exchanges/bybit/bybit_websocket.go

```diff
@@ -0,0 +1,548 @@
+package bybit
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/gorilla/websocket"
+	"github.com/thrasher-corp/gocryptotrader/common"
+	"github.com/thrasher-corp/gocryptotrader/common/crypto"
+	"github.com/thrasher-corp/gocryptotrader/currency"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/asset"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/order"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/orderbook"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/stream"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/ticker"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/trade"
+	"github.com/thrasher-corp/gocryptotrader/log"
+)
+
+const (
+	bybitWSBaseURL      = "wss://stream.bybit.com/"
+	wsSpotPublicTopicV2 = "spot/quote/ws/v2"
+	wsSpotPrivate       = "spot/ws"
+	bybitWebsocketTimer = 20 * time.Second
+	wsOrderbook         = "depth"
+	wsTicker            = "bookTicker"
+	wsTrades            = "trade"
+	wsKlines            = "kline"
+
+	wsAccountInfo    = "outboundAccountInfo"
+	wsOrderExecution = "executionReport"
+	wsTickerInfo     = "ticketInfo"
+
+	sub    = "sub"    // event for subscribe
+	cancel = "cancel" // event for unsubscribe
+)
+
+var comms = make(chan stream.Response)
+
+// WsConnect connects to a websocket feed
+func (by *Bybit) WsConnect() error {
+	if !by.Websocket.IsEnabled() || !by.IsEnabled() || !by.IsAssetWebsocketSupported(asset.Spot) {
+		return errors.New(stream.WebsocketNotEnabled)
+	}
+	var dialer websocket.Dialer
+	err := by.Websocket.Conn.Dial(&dialer, http.Header{})
+	if err != nil {
+		return err
+	}
+	by.Websocket.Conn.SetupPingHandler(stream.PingHandler{
+		MessageType: websocket.TextMessage,
+		Message:     []byte(`{"op":"ping"}`),
+		Delay:       bybitWebsocketTimer,
+	})
+
+	by.Websocket.Wg.Add(1)
+	go by.wsReadData(by.Websocket.Conn)
+	if by.IsWebsocketAuthenticationSupported() {
+		err = by.WsAuth(context.TODO())
+		if err != nil {
+			by.Websocket.DataHandler <- err
+			by.Websocket.SetCanUseAuthenticatedEndpoints(false)
+		}
+	}
+
+	by.Websocket.Wg.Add(1)
+	go by.WsDataHandler()
+	return nil
+}
+
+// WsAuth sends an authentication message to receive auth data
+func (by *Bybit) WsAuth(ctx context.Context) error {
+	var dialer websocket.Dialer
+	err := by.Websocket.AuthConn.Dial(&dialer, http.Header{})
+	if err != nil {
+		return err
+	}
+	by.Websocket.Wg.Add(1)
```

Comment:
Please add in a pinghandler for the auth connection as well. I am getting disconnection errors without it.
```go
by.Websocket.AuthConn.SetupPingHandler(stream.PingHandler{
		MessageType: websocket.TextMessage,
		Message:     []byte(`{"op":"ping"}`),
		Delay:       bybitWebsocketTimer,
	})
```

Reply:
Okay, I missed that. Sorry. I have added it now. Thanks for pointing it out. @gloriousCode